### PR TITLE
Use kad.RoutingTable, kad.NodeID and kad.NodeAddr interfaces

### PR DESCRIPTION
**Pull request by dennis-tra**
Refactor code to use the new kad.RoutingTable, kad.NodeID and kad.NodeAddr interfaces.
Consolidate libp2p and ipfs-specific implementations into a new libp2p package.

### CODE REVIEW

File: kad/kad.go [hunk 1]
```diff
@@ -82,15 +82,50 @@ type NodeID[K Key[K]] interface {
...
+type Message interface{}
+
+type Request[K Key[K], A Address[A]] interface {
+	Message
+
+	// Target returns the target key and true, or false if no target key has been specfied.
+	Target() K
+
+	// EmptyResponse returns an empty response struct for this request message
+	// TODO: this is a weird patter, let's try to remove this.
```
Comment (iand) [inline kad/kad.go:123]: It's hopefully not needed in the future

File: network/endpoint/endpoint.go [hunk 2]
```diff
@@ -4,62 +4,79 @@ import (
...
+	Connectedness(kad.NodeID[K]) (Connectedness, error)
 }
 
 // StreamID is a unique identifier for a stream.
 type StreamID uint64
 
 // SimEndpoint is a simulated endpoint that doesn't operate on real network
-type SimEndpoint[K kad.Key[K]] interface {
-	ServerEndpoint[K]
+type SimEndpoint[K kad.Key[K], A kad.Address[A]] interface {
```
Comment (iand) [inline network/endpoint/endpoint.go:77]: Can we remove this or move to `sim`?

File: network/endpoint/endpoint.go [hunk 3]
```diff
@@ -4,62 +4,79 @@ import (
...
-type ResponseHandlerFn[K kad.Key[K]] func(context.Context, message.MinKadResponseMessage[K], error)
+type ResponseHandlerFn[K kad.Key[K], A kad.Address[A]] func(context.Context, kad.Response[K, A], error)
 
 // Endpoint defines how Kademlia nodes interacts with each other.
-type Endpoint[K kad.Key[K]] interface {
+type Endpoint[K kad.Key[K], A kad.Address[A]] interface {
 	// MaybeAddToPeerstore adds the given address to the peerstore if it is
 	// valid and if it is not already there.
-	MaybeAddToPeerstore(context.Context, address.NodeAddr[K], time.Duration) error
+	MaybeAddToPeerstore(context.Context, kad.NodeInfo[K, A], time.Duration) error
```
Comment (iand) [inline network/endpoint/endpoint.go:42]: I wonder if `MaybeAddToPeerstore` and `NetworkAddress` actually belong on `NetworkedEndpoint`

File: network/endpoint/endpoint.go [hunk 4]
```diff
@@ -4,62 +4,79 @@ import (
...
 	// the response with the given handler.
-	SendRequestHandleResponse(context.Context, address.ProtocolID, address.NodeID[K],
-		message.MinKadMessage, message.MinKadMessage, time.Duration,
-		ResponseHandlerFn[K]) error
+	SendRequestHandleResponse(context.Context, address.ProtocolID, kad.NodeID[K],
+		kad.Message, kad.Message, time.Duration,
+		ResponseHandlerFn[K, A]) error
 
 	// KadKey returns the KadKey of the local node.
 	KadKey() K
```
Comment (iand) [inline network/endpoint/endpoint.go:50]: Let's rename to `Key`

File: sim/message.go [hunk 5]
```diff
@@ -2,37 +2,35 @@ package sim
 
 import (
 	"github.com/plprobelab/go-kademlia/kad"
-	"github.com/plprobelab/go-kademlia/network/address"
-	"github.com/plprobelab/go-kademlia/network/message"
 )
 
-// Message is a simple implementation of `MinKadRequestMessage` and `MinKadResponseMessage`.
+// SimMessage is a simple implementation of `Request` and `Response`.
```
[the remark this answers is not in the document]
Comment (dennis-tra) [reply sim/message.go:7]: ah, I moved this type around and forgot to give it the original name when I moved it back 👍 will change back

File: network/endpoint/endpoint.go (hunk 2 above)
Comment (dennis-tra) [reply network/endpoint/endpoint.go:77]: 👍

File: network/endpoint/endpoint.go (hunk 3 above)
Comment (dennis-tra) [reply network/endpoint/endpoint.go:42]: I'm not as deep into the Endpoint abstraction to really judge :/

**Pull request merged**